jsonrpc/filter_manager.go

```diff
@@ -15,57 +16,61 @@ import (
 	"github.com/hashicorp/go-hclog"
 )
 
-type Filter struct {
-	id string
+var ErrFilterDoesNotExists = errors.New("filter does not exists")
 
-	// block filter
-	block *headElem
+// defaultTimeout is the timeout to remove the filters that doesn't have web socket stream
+var defaultTimeout = 1 * time.Minute
+
+// filter is an interface that BlockFilter and LogFilter implement
+type filter interface {
+	// isWS returns the flag indicating the filter has web socket stream
+	isWS() bool
 
-	// log cache
-	logs []*Log
+	// getFilterBase returns filterBase that has common fields
+	getFilterBase() *filterBase
+
+	// getUpdates returns stored data in string
+	getUpdates() (string, error)
+
+	// sendUpdates write stored data to web socket stream
+	sendUpdates() error
+}
 
-	// log filter
-	logFilter *LogFilter
+// filterBase is a struct for common fields between BlockFilter and LogFilter
+type filterBase struct {
+	// UUID, a key of filter for client
+	id string
 
-	// index of the filter in the timer array
-	index int
+	// index in the timeouts heap, -1 for non-existing index
+	heapIndex int
 
-	// next time to timeout
-	timestamp time.Time
+	// timestamp to be expired
+	expiredAt time.Time
 
 	// websocket connection
 	ws wsConn
 }
 
-func (f *Filter) getFilterUpdates() (string, error) {
-	if f.isBlockFilter() {
-		// block filter
-		headers, newHead := f.block.getUpdates()
-		f.block = newHead
-
-		updates := []string{}
-		for _, header := range headers {
-			updates = append(updates, header.Hash.String())
-		}
-
-		return fmt.Sprintf("[\"%s\"]", strings.Join(updates, "\",\"")), nil
-	}
-	// log filter
-	res, err := json.Marshal(f.logs)
-	if err != nil {
-		return "", err
+// newFilterBase initializes filterBase with unique ID
+func newFilterBase(ws wsConn) filterBase {
+	return filterBase{
+		id:        uuid.New().String(),
+		ws:        ws,
+		heapIndex: -1,
```

Comment:
Can we extract out the -1 to a constant?

Reply:
Added at b0728d05324cea09366d9761c7661205e4e6dcac